File: Fortran/IBM/execute/F2003/cmdline/Misc/fxclms05.c
#include<stdio.h>
#include<unistd.h>



int main(int argc, char ** argv, char ** envp)
{
/*

The main idea for this test is to form maximum length of a command line.
Different os has defferent length limit.
To issue a command, we can call execlp that will be independent from shell
To determine the maximum possible length, we should calculate
 
    limit - the size used for environment variables

The command format formed is:
  
    fxclms050 OptionLength xx.......x 

*/

  const char* p1="fxclms050";
  char  para2[500000];
  char  c; 
  int   rc;
  int   i;
  int   cnt;
  int   length;
  char  para1[8];

  while (*envp != 0)
  {
    cnt += strlen(*envp) + 1;
    envp++;
  }
 
/* 
 AIX:   /usr/include/sys/limits.h      24k = 24576
 MACOS: /usr/include/sys/syslimits.h   64k = 65536
 SLES:  /usr/include/linux/limits.h    128k= 131072
 RHEL:  /usr/include/linux/limits.h    128k= 131072

 isAIX:
     0)  OS="AIX"
     1)  OS="LINUX"
     2)  OS="MACOSX"
*/
  switch (atoi(*(++argv)))
  {
   case 0:
     length = 24576;
     break;
   case 1:
     length = 131072; 
     break;
   case 2:
     length = 65536; 
     break;
   default:
     printf("Unknown OS!\n");
     return 11;
  }

  cnt = length - cnt - (strlen(p1) +1 ) - 8;
  sprintf(para1, "%d", cnt);

  for (i=0; i < cnt; i++)
  {
    para2[i] = 'x';
  }
  para2[cnt] = '\0';

  rc = execlp(p1, p1, para1, para2, (char *)0);

  printf("ret is %d\n", rc );

}

File: Fortran/IBM/execute/F2003/cmdline/Misc/fxclms05.f
! *********************************************************************
!*  ===================================================================
!*
!*  DATE                       : Oct 1, 2003
!*
!*  PRIMARY FUNCTIONS TESTED   	: COMMAND_ARGUMENT_COUNT()
!*                            	: GET_COMMAND(COMMAND, LENGTH, STATUS)
!*                            	: GET_COMMAND_ARGUMENT(NUMBER, VALUE, LENGTH, STATUS)
!*                             	: GET_ENVIRONMENT_VARIABLE(NAME, VALUE, LENGTH, STATUS, TRIM_NAME)
!*
!*  SECONDARY FUNCTIONS TESTED :
!*
!*  REFERENCE                  : Feature 252525
!*
!*  REQUIRED COMPILER OPTIONS  :
!*
!*  KEYWORD(S)                 :
!*  TARGET(S)                  :
!*  NUMBER OF TESTS CONDITIONS :
!*
!*  DESCRIPTION                : Tests command line intrinsic routines by passing a long command line
!*                             : through a system call to imitate issuing a command directly on
!*                             : command line without involving a shell
!*                             : fxclms05.c forms a maximum length of a command with the form :
!*                             : fxclms050 OptionLength xx.......x
!*                             : Then the c program call a system call "execlp" to execute the formed
!*                             : command line.
!*                             : The detail explanation on the forming of a command line is in the c
!*                             : program "fxclms05.c"
!*                             : The maximum lengths (ARG_MAX) on various systems :
!*
!*                             : AIX:   /usr/include/sys/limits.h      24k = 24576
!*                             : MACOS: /usr/include/sys/syslimits.h   64k = 65536
!*                             : SLES:  /usr/include/linux/limits.h    128k= 131072
!*                             : RHEL:  /usr/include/linux/limits.h    128k= 131072
!*                             : Call chain:
!*                             : fxclms05.sh => fxclms05.c => fxclms050.f
!234567890123456789012345678901234567890123456789012345678901234567890

